index.bs

```diff
@@ -2055,86 +2055,98 @@ for the APIs that they design.
-is more important when the algorithms would be more repetitive.
-These formal specification techniques can include dependencies on
-languages such as [[WEBIDL]], on grammar productions,
-or the development of new languages for defining pieces of the specification.
-Merely having a formal syntax is not sufficient, though;
-the requirements that the formal definitions place on implementations must still be clearly defined,
-which generally requires specifying some algorithms (or, sometimes, another layer of formal syntax) that define what the formal syntax means,
-and then using the more-readable formal syntax to avoid
-excessive repetition of the less-readable algorithms.
-There are also possibilities in-between, for example,
-if it is desirable to ensure that an algorithm can be implemented with a state machine,
-it may be preferable to specify it in an algorithmic way that incorporates explicit states
-so that the specification authors avoid accidentally creating more states than intended.
+Write algorithms in a way that is clear and concise.
+
+The most common way to write algorithms
+is to write an explicit sequence of steps.
+This often looks like pseudo-code.
+
+<div class="example">
+The {{HTMLDialogElement/showModal()}} method is described as a numbered sequence of steps
+which clearly explains when to throw exceptions
+and when to run algorithms defined in other parts of the HTML spec.
+</div>
+
+When writing a sequence of steps,
+imagine that it is a piece of functional code.
+
+- Clearly specify the inputs and outputs,
+name the algorithm and the variables it uses well,
+and explicitly note the points in the algorithm where the algorithm may
+return a result or error.
+- As much as possible, avoid writing algorithms which have side effects.
+
+Summarize the purpose of the algorithm before going into detail,
+so that readers can decide whether to read the steps or skip over them.
+For example take the following steps, which ensure that there is at most one pending X
+callback per toplevel browsing context".
+
+Issue: The paragraph below may not be comprehensive enough.
+
+A plain sequence of steps is not always the best way to write an algorithm.
+For example, it might make sense to define
+or re-use a formal syntax or grammar
+to avoid repetition,
+or define specific states to be used in a state machine.
+When using extra constructs like these,
+the earlier advice still applies.
+
+Issue: I don't understand the paragraph below.
 
 When specifying in an algorithmic way,
 it is generally good to match the algorithms that implementations would use,
 even if this is somewhat less convenient for specification authors,
-as long as doing so would not significantly increase the specification's complexity.
-For example, it is better to specify CSS selector matching
+as long as doing so wouldn't significantly increase the specification's complexity.
+For example, it's better to specify CSS selector matching
```

Comment:
https://drafts.csswg.org/selectors-4/#match-against-element
goes right-to-left.

Selector matching is done in two ways - during style, you start with an element, and go from right to left, and during querySelector, when you would go left to right...